### PR TITLE
Run a callback on each test failure

### PR DESCRIPTION
**Pull request by julien-meichelbeck**
Sometimes you want to clean up stuff between two retries (clean db, redis, refresh stuff).

### CODE REVIEW

File: lib/minitest/retry.rb [hunk 1]
```diff
@@ -32,11 +41,11 @@ def self.failure_to_retry?(failures = [])
 
     module ClassMethods
       def run_one_method(klass, method_name)
-        retry_count = Minitest::Retry.retry_count
         result = super(klass, method_name)
         return result unless Minitest::Retry.failure_to_retry?(result.failures)
+        Minitest::Retry.failure_callback.call if Minitest::Retry.failure_callback
```
Comment (y-yagi) [inline lib/minitest/retry.rb]: I think that this process should not be executed in the case of skip. 
Therefore, I think that it is better to do after `!result.skipped?`.

File: test/minitest/retry_test.rb [hunk 2]
```diff
@@ -186,5 +186,42 @@ def raise_test_error
     end
   end
 
+  def test_run_failure_callback_on_failure
+    on_failure_block_has_ran = false
+    output = capture_stdout do
```
Comment (y-yagi) [inline test/minitest/retry_test.rb]: Looks like `output` variable is not used. Please remove.

File: test/minitest/retry_test.rb [hunk 3]
```diff
@@ -186,5 +186,42 @@ def raise_test_error
...
+  def test_do_not_run_failure_callback_on_success
+    on_failure_block_has_ran = false
+    output = capture_stdout do
+      retry_test = Class.new(Minitest::Test) do
+        Minitest::Retry.use!
+        Minitest::Retry.on_failure do
+          on_failure_block_has_ran = true
+        end
+
+        def fail
```
Comment (y-yagi) [inline test/minitest/retry_test.rb]: This test is not *fail*. So I think `success` is better name.

**Comment by y-yagi** [comment]
Thanks for your PR! I like this approach. Please check my comments :wink:

**Comment by julien-meichelbeck** [comment]
Cleaned up my mess and added the new tests in another test file, this one is starting to get quite long (hope you don't mind)!

**Comment by y-yagi** [comment]
Thanks!  But I like to leave the test file as one. I prefer the first approach, is it possible to revert it? (If you allow, I will revert :smile_cat: )

**Comment by julien-meichelbeck** [comment]
Sure please do !

**Comment by y-yagi** [comment]
I merged locally.  Thank you for your contribution!!

**Comment by y-yagi** [comment]
Released as v0.1.7

**Pull request merged**